services/auth/source/ldap/source_authenticate.go

```diff
@@ -20,6 +20,9 @@ import (
 // Authenticate queries if login/password is valid against the LDAP directory pool,
 // and create a local user if success when enabled.
 func (source *Source) Authenticate(user *user_model.User, userName, password string) (*user_model.User, error) {
+	if user != nil {
```

Comment:
I think userName should not be empty? It should be come from inputbox.

Reply:
But the username may be different from the loginname that must be used with ldap. It could also be an email address. 

In ldap you want to search for the login name only (this was at least what was working up to version 1.15.9).

Reply:
So that should be changed in the progress of LDAP users sync but not when authenticating. When sync a user, we need to know should we use login name as identify or use email?

Reply:
I still think this is the appropriate place - when you login using your gitea username or email address, validation must be done using the loginname that was stored for the account. 

I improved wording a bit in a new commit.